File: Java/No05.java
/**
 * 输入一个链表的头节点，从尾到头打印每个节点的值。
 */

public class No05 {

    public static void main(String[] args) {
        Node node1 = new Node("A");
        Node node2 = new Node("B");
        Node node3 = new Node("C");
        Node node4 = new Node("D");
        Node node5 = new Node("E");
        node1.setNext(node2);
        node2.setNext(node3);
        node3.setNext(node4);
        node4.setNext(node5);
        Node newNode = reverse2(node1);
        while (newNode != null) {
            System.out.print(newNode.data + " ");
            newNode = newNode.getNext();
        }


    }

    private static Node reverse(Node head) {
        if (head.next == null) {
            return head;
        }
        Node reverseHead = reverse(head.getNext());
        head.getNext().setNext(head);
        head.setNext(null);

        return reverseHead;

    }

    private static Node reverse2(Node head) {
        Node pre = head;
        Node cur = head.getNext();
        Node temp;
        while (cur != null) {
            temp = cur.getNext();
            cur.setNext(pre);
            pre = cur;
            cur = temp;
        }
        head.setNext(null);
        return pre;
    }
}

class Node {
    String data;
    Node next;

    public Node(String data) {
        super();
        this.data = data;
    }

    public Node(String data, Node next) {
        super();
        this.data = data;
        this.next = next;
    }

    public String getData() {
        return data;
    }

    public void setData(String data) {
        this.data = data;
    }

    public Node getNext() {
        return next;
    }

    public void setNext(Node next) {
        this.next = next;
    }


}

File: Kotlin/No05.kt
class No05 {

    fun main() {
        val node1 = Node("A")
        val node2 = Node("B")
        val node3 = Node("C")
        val node4 = Node("D")
        val node5 = Node("E")
        node1.next = node2
        node2.next = node3
        node3.next = node4
        node4.next = node5

        var newNode: Node? = reverseNode1(node1)
        // var newNode: Node? = reseverNode2(node1)
        while (newNode != null) {
            print(newNode.data)
            newNode = newNode.next
        }
    }

    fun reverseNode1(node: Node): Node {
        if (node.next == null) {
            return node
        }
        val reverseHead = reverseNode1(node.next!!)
        node.next!!.next = node
        node.next = null
        return reverseHead
    }

    fun reseverNode2(node: Node): Node {
        var pre = node
        var cur = pre.next
        pre.next = null
        var temp: Node? = null
        while (cur != null) {
            temp = cur.next
            cur.next = pre
            pre = cur
            cur = temp
        }
        return pre
    }
}
